homeassistant/components/stream/__init__.py

```diff
@@ -0,0 +1,156 @@
+        vol.Optional(CONF_KEEPALIVE, default=False): cv.boolean,
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+
+# pylint: disable=dangerous-default-value
+@bind_hass
+async def async_request_stream(hass, stream_source, options={}, preload=False):
+    """Set up stream with token."""
+    if DOMAIN not in hass.config.components:
+        raise HomeAssistantError("Stream is not configured.")
+
+    if not stream_source:
+        raise HomeAssistantError("Invalid stream source.")
+
+    keepalive = hass.data[DOMAIN][CONF_KEEPALIVE]
+    try:
+        streams = hass.data[DOMAIN][ATTR_STREAMS]
+        stream = streams.get(stream_source)
+        if not stream:
+            stream = Stream(hass, stream_source,
+                            options=options, keepalive=keepalive)
+            hass.data[DOMAIN][ATTR_STREAMS][stream_source] = stream
+        if not preload and not stream.access_token:
+            stream.access_token = generate_secret()
+            stream.start()
+        return stream.access_token
+    except Exception:
+        raise HomeAssistantError('Unable to get stream')
+
+
+@bind_hass
+def get_stream(hass, stream_source):
+    """Return available stream."""
+    streams = hass.data[DOMAIN][ATTR_STREAMS]
+    return streams.get(stream_source)
+
+
+@bind_hass
+def get_url(hass, fmt, stream_source=None, token=None):
+    """Return stream access url."""
+    if not token:
+        stream = hass.data[DOMAIN][ATTR_STREAMS].get(stream_source)
+        if stream:
+            token = stream.access_token
+    if token:
+        return hass.data[DOMAIN][ATTR_ENDPOINTS][fmt].format(
+            hass.config.api.base_url, token)
+    return None
+
+
+async def async_setup(hass, config):
+    """Set up stream."""
+    conf = config[DOMAIN] if config.get(DOMAIN, {}) else {}
+
+    hass.data[DOMAIN] = {}
+    hass.data[DOMAIN][CONF_KEEPALIVE] = conf.get(CONF_KEEPALIVE)
+    hass.data[DOMAIN][ATTR_ENDPOINTS] = {}
+    hass.data[DOMAIN][ATTR_STREAMS] = {}
+
+    # Setup HLS
+    hls_endpoint = await async_setup_hls(hass)
+    hass.data[DOMAIN][ATTR_ENDPOINTS]['hls'] = hls_endpoint
+
+    return True
+
+
+class Stream:
+    """Represents a single stream."""
+
+    # pylint: disable=dangerous-default-value
+    def __init__(self, hass, source, options={}, keepalive=False):
+        """Initialize a stream."""
+        self.hass = hass
+        self.source = source
+        self.options = options
+        self.keepalive = keepalive
+        self.access_token = None
+        self.__container = None
+        self.__thread = None
+        self.__thread_quit = None
+        self.__outputs = {}
+
+    @property
+    def container(self):
+        """Return container."""
+        return self.__container
+
+    @property
+    def outputs(self):
+        """Return stream outputs."""
+        return self.__outputs
+
+    def add_provider(self, provider):
+        """Add provider output stream."""
+        if not self.__outputs.get(provider.format):
```

Comment:
Shall this method replace the existing provider by new provider with same format instead of return old one?

Reply:
No, because then you would lose already buffered stream segments.